glom/streaming.py

```diff
@@ -0,0 +1,196 @@
+"""
+Helpers for streaming use cases -- that is, specifier types which yield their
+results incrementally so that they can be applied to targets which
+are themselves streaming (e.g. chunks of rows from a database, lines
+from a file) without excessive memory usage.
+"""
+from __future__ import unicode_literals
+
+from itertools import islice, dropwhile
+from functools import partial
+try:
+    from itertools import izip, izip_longest
+except ImportError:
+    izip = zip  # py3
+    from itertools import zip_longest as izip_longest
+
+from boltons.iterutils import split_iter, chunked_iter, windowed_iter, unique_iter
+
+from .core import glom, T, STOP, SKIP, Check, _MISSING, Path, TargetRegistry, Call, Spec, S
+from .reduction import Flatten
+
+"""
+itertools to add:
+
+(Iter(T.items)
+ .filter(T.is_activated)
+ .map(fetch_from_db))
+
+(Iter((T.items, Check(T.is_activated, default=SKIP), fetch_from_db)))
+
+# should there be a .first() or something?
+"""
+
+class Iter(object):
+    """``Iter()`` is glom's counterpart to the built-in :func:`iter()`
+    function. Given an iterable target, yields the result of applying
+    the passed spec to each element of the target. Basically, a lazy
+    version of the default list-spec behavior.
+
+    ``Iter()`` also respects glom's :data:`~glom.SKIP` and
+    :data:`~glom.STOP` singletons for filtering and breaking
+    iteration.
+
+    Args:
+
+       subspec: A subspec to be applied on each element from the iterable.
+       sentinel: Keyword-only argument, which, when found in the
+         iterable stream, causes the iteration to stop. Same as with the
+         built-in :func:`iter`.
+
+    >>> glom(['1', '2', '1', '3'], (Iter(int), set, tuple))
+    (1, 2, 3)
+
+
+    """
+    def __init__(self, subspec=T, **kwargs):
+        self.subspec = subspec
+        self._spec_stack = kwargs.pop('spec_stack', [])
+
+        self.sentinel = kwargs.pop('sentinel', STOP)
+        if kwargs:
+            raise TypeError('unexpected keyword arguments: %r' % sorted(kwargs))
+        return
```

Comment:
I see... this `_spec_stack` is kind of an alternative to having a linked list of these guys in memory